### ISSUE TITLE
Dynamic menus break after visiting page with ngx-quill

### ISSUE THREAD

**Issue by TWinters** [KillerCodeMonkey/ngx-quill#324]
Hi there!

Great work on this package, I really enjoy using this component!

However, I ran into an odd issue on multiple occassions, and had a hard time figuring out what is causing this, and I think it might be a bug in ngx-quill somehow.
I set up a repository reproducing the issue [here](https://github.com/TWinters/ngx-quill-click-bug-example).

The issue is that when having routing links dynamically loaded in like this: 
```
<div *ngFor="let item of getMenuItems()">
    <a [routerLink]="item.route">
        {{item.text}}
    </a>
</div>
```
The anchor elements will work, until you visit a page with a ngx-quill editor on. Then they will not route you to pages any more, until you refresh the page. The exact same happens with dynamically loaded elements with `(click)`: these will not get executed.

In the linked repository, this bug occurs in a nav-bar component that stays present. However, I previously also encountered this behaviour when going to another page that loaded clickable or routerLink elements after having visited a page with a quill editor.

Do you think this is a bug caused by ngx-quill? Do you know a way to fix this issue?

**Comment by KillerCodeMonkey** [comment]
Maybe you get an Error in your Javascript console? And this Error Breaks
your App?



Am Fr., 25. Jan. 2019, 23:02 hat Thomas Winters <notifications@github.com>
geschrieben:

**Comment by TWinters** [comment]
Thanks for the quick reply!

I don't get an error, but I should have indeed mentioned that I get a warning in my Javascript console:
`zone.js:1883 [Violation] Added synchronous DOM mutation listener to a 'DOMNodeInserted' event. Consider using MutationObserver to make the page more responsive.` in Google Chrome and `Use of Mutation Events is deprecated. Use MutationObserver instead.` in Firefox.
Do you think this might be causing the issue? My gut feeling is that there might be some script loaded in somewhere that receives click events and cancels them somehow, although that wouldn't explain why elements with a routeLink that are hardcoded in HTML are not affected.

The error I reproduced in [the repository](https://github.com/TWinters/ngx-quill-click-bug-example) is currently indeed breaking my app, and causing me not to be able to push a certain update. I wanted to have quite radically different modes of navigation when the user is on mobile, and created an easy and clean way to update both navigation systems by making them share a list of objects specifying the menu items. However, now the whole navigation system breaks after visiting a page where a `quill-editor` component has been loaded, meaning the user is stuck until he both succeeds going to a page without a quill-editor and refreshes the page.
My only current alternative is hard coding the menus again by copy pasting all of my menu items code for every menu item, and do this for both styles of menu, which makes them all harder to cleanly update later.

**Comment by KillerCodeMonkey** [comment]
seems like it has something to do, that your routings are returned by a function `getMenuItems()`.

When i store them on a static class property and use ngFor in the template it works.

So i think, because of the initialisation of ngx-quill component the changedetection triggers and there gets something messed up in angular and its routing.

**Comment by KillerCodeMonkey** [comment]
Feel free to let me know when you get IT working or get nee information. In
General i so Not think it is a ngx-quill issue. Because i do not do
anything with Routing and it working in all Other cases.

But Strange Thing thou.

Could you Check what Happens, when you use HashLocationStrategy?

Am Sa., 26. Jan. 2019, 13:21 hat Thomas Winters <notifications@github.com>
geschrieben:

**Comment by KillerCodeMonkey** [comment]
A short Idea. Try to Change Changedetection strategy to onPush for your
navbar component

Am Sa., 26. Jan. 2019, 13:31 hat Bengt Weiße <bengtler@googlemail.com>
geschrieben:

**Comment by TWinters** [comment]
Changing to HashLocationStrategy didn't fix it, but using your ChangeDetection fix does indeed work!
Using the OnPush change detection however gives rise to some issues in my app (e.g. the navbar not updating when the user is logged in), but I can probably work these out by reading more on how OnPush change detection method is best used. Thanks!

Does this mean that the ngx-quill component in certain cases fires a lot of (unnecessary) changes? I have had similar issues when creating a preview of the outputted quill text when the text contained a (video) iframe, because the iframe would reload on page scrolls or output text clicks (only when quill editor on same page). I also have pages that give an warning if you leave with unsaved changes, that sometimes warn when nothing was changed, but only have this issue on pages with a quill editor. These are of course very minor issues which I didn't feel the need to report, as I am not 100% sure they're caused by ngx-quill, but they might be relevant here. Feel free to reopen the issue or contact me if you'd like me to elaborate on these.

Anyway, I'm now using your suggestion with static class attributes instead of getter methods, and now know that I can try OnPush change detection methods in case I have similar issues later. Thank you very much for solving my issue! :)

**Comment by KillerCodeMonkey** [comment]
This has Something to do with angulars ChangeDetection i think.

In my example everything works flawless

I do Not know, why IT gets so crazy with Routing.

Every click and scroll can Trigger angulars Change Detection.

I learned that calling functions from the Templates should be avoided If
you are using fixed values or only need to calculate them on Special
conditions.

Your nav component should have an @Input to Switch for example between
loggedin and Not loggedin nav items and so on.



Am 26.01.2019 20:38 schrieb "Thomas Winters" <notifications@github.com>:

Closed #324 <https://github.com/KillerCodeMonkey/ngx-quill/issues/324>.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/KillerCodeMonkey/ngx-quill/issues/324#event-2098990428>,
or mute the thread

**Comment by TWinters** [comment]
Thank you for the helpful advice! The navbar in my app now works flawlessly: it indeed needed the user as an input to the navbar component instead of pulling that variable from the authentication service.

Thank you very much for the tips and of course for the this amazing library! :D
